rust-runtime/aws-smithy-types/src/body/http_body_1_x.rs

```diff
@@ -0,0 +1,269 @@
+ */
+
+//! Adapters to use http-body 1.0 bodies with SdkBody & ByteStream
+
+use std::pin::Pin;
+use std::task::{ready, Context, Poll};
+
+use bytes::Bytes;
+use http_body_util::BodyExt;
+use pin_project_lite::pin_project;
+
+use crate::body::{Error, SdkBody};
+
+impl SdkBody {
+    /// Construct an `SdkBody` from a type that implements [`http_body_1_0::Body<Data = Bytes>`](http_body_1_0::Body).
+    ///
+    /// _Note: This is only available with `http-body-1-0` enabled._
+    pub fn from_body_1_x<T, E>(body: T) -> Self
+    where
+        T: http_body_1_0::Body<Data = Bytes, Error = E> + Send + Sync + 'static,
+        E: Into<Error> + 'static,
+    {
+        SdkBody::from_body_0_4_internal(Http1toHttp04::new(body.map_err(Into::into)))
+    }
+}
+
+pin_project! {
+    struct Http1toHttp04<B> {
+        #[pin]
+        inner: B,
+        trailers: Option<http_1x::HeaderMap>,
+    }
+}
+
+impl<B> Http1toHttp04<B> {
+    fn new(inner: B) -> Self {
+        Self {
+            inner,
+            trailers: None,
+        }
+    }
+}
+
+impl<B> http_body_0_4::Body for Http1toHttp04<B>
+where
+    B: http_body_1_0::Body,
+{
+    type Data = B::Data;
+    type Error = B::Error;
+
+    fn poll_data(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<Self::Data, Self::Error>>> {
+        loop {
+            let this = self.as_mut().project();
+            match ready!(this.inner.poll_frame(cx)) {
+                Some(Ok(frame)) => {
+                    let frame = match frame.into_data() {
+                        Ok(data) => return Poll::Ready(Some(Ok(data))),
+                        Err(frame) => frame,
+                    };
+                    // when we get a trailers frame, store the trailers for the next poll
+                    if let Ok(trailers) = frame.into_trailers() {
+                        this.trailers.replace(trailers);
+                        return Poll::Ready(None);
+                    };
+                    // if the frame type was unknown, discard it. the next one might be something
+                    // useful
+                }
+                Some(Err(e)) => return Poll::Ready(Some(Err(e))),
+                None => return Poll::Ready(None),
+            }
+        }
+    }
+
+    fn poll_trailers(
+        self: Pin<&mut Self>,
+        _cx: &mut Context<'_>,
+    ) -> Poll<Result<Option<http::HeaderMap>, Self::Error>> {
+        // all of the polling happens in poll_data, once we get to the trailers we've actually
+        // already read everything
+        let this = self.project();
+        match this.trailers.take() {
+            Some(headers) => Poll::Ready(Ok(Some(convert_header_map(headers)))),
+            None => Poll::Ready(Ok(None)),
+        }
+    }
+
+    fn is_end_stream(&self) -> bool {
+        self.inner.is_end_stream()
+    }
+
+    fn size_hint(&self) -> http_body_0_4::SizeHint {
+        let mut size_hint = http_body_0_4::SizeHint::new();
```

Comment:
I assume we use `http_body_0_4::SizeHint` directly here as opposed to first using `http_body_1_0::SizeHint` and converting what that returns to `http_body_0_4::SizeHint` ?